pkg/repository/v1manifest/repo.go

```diff
@@ -113,43 +202,84 @@ func NewTimestamp(initTime time.Time) *Timestamp {
 	}
 }
 
-// SignAndWrite creates a manifest and writes it to out.
-func SignAndWrite(out io.Writer, role ValidManifest, keyID string, privKey *crypto.RSAPrivKey) error {
-	payload, err := cjson.Marshal(role)
-	if err != nil {
-		return err
+// NewComponent creates a Component object
+func NewComponent(id, name, desc string, initTime time.Time) *Component {
+	return &Component{
+		SignedBase: SignedBase{
+			Ty:          ManifestTypeComponent,
+			SpecVersion: CurrentSpecVersion,
+			Expires:     initTime.Add(ManifestsConfig[ManifestTypeComponent].Expire).Format(time.RFC3339),
+			Version:     1,
+		},
+		ID:          id,
+		Name:        name,
+		Description: desc,
+		Platforms:   make(map[string]map[string]VersionItem),
 	}
+}
 
-	sign, err := privKey.Signature(payload)
+// NewKeyInfo creates a KeyInfo object
+func NewKeyInfo(priv crypto.PrivKey) (*KeyInfo, error) {
+	pubBytes, err := priv.Public().Serialize()
 	if err != nil {
-		return errors.Trace(err)
+		return nil, err
 	}
+	return &KeyInfo{
+		Algorithms: []string{"sha256"},
+		Type:       "rsa",
+		Value: map[string]string{
+			"public": string(pubBytes),
+		},
+		Scheme: "rsassa-pss-sha256",
+	}, nil
+}
 
-	manifest := Manifest{
-		Signatures: []signature{{
-			KeyID: keyID,
-			Sig:   sign,
-		}},
-		Signed: role,
+// ID returns the SH256 hash of the key
+func (k *KeyInfo) ID() (string, error) {
+	info, err := cjson.Marshal(k)
+	if err != nil {
+		return "", err
 	}
-
-	encoder := json.NewEncoder(out)
-	encoder.SetIndent("", "\t")
-	return encoder.Encode(manifest)
+	hash := sha256.Sum256(info)
+	return hex.EncodeToString(hash[:]), nil
 }
 
 // SetVersions sets file versions to the snapshot
-func (manifest *Snapshot) SetVersions(manifestList map[string]ValidManifest) *Snapshot {
+func (manifest *Snapshot) SetVersions(manifestList map[string]ValidManifest) (*Snapshot, error) {
 	if manifest.Meta == nil {
 		manifest.Meta = make(map[string]FileVersion)
 	}
 	for _, m := range manifestList {
+		bytes, err := cjson.Marshal(m)
+		if err != nil {
+			return nil, err
+		}
 		manifest.Meta[m.Filename()] = FileVersion{
 			Version: m.Base().Version,
-			// TODO length
+			Length:  uint(len(bytes)),
```

Comment:
the length should be the length of the whole manifest, not just the signed field